openjdk/mmtkBarrierSetAssembler_x86.cpp

```diff
@@ -112,35 +112,32 @@ void MMTkBarrierSetAssembler::eden_allocate(MacroAssembler* masm, Register threa
     __ movptr(cursor, end);
 
 #ifdef MMTK_ENABLE_GLOBAL_ALLOC_BIT
-    Register tmp3 = rscratch2;
+    Register tmp3 = rdi;
     Register tmp2 = rscratch1;
-    Register tmp4 = t1;
-    assert_different_registers(obj, tmp2, tmp3, tmp4, rcx);
+    assert_different_registers(obj, tmp2, tmp3, rcx);
 
     // tmp2 = load-byte (SIDE_METADATA_BASE_ADDRESS + (obj >> 6));
     __ movptr(tmp3, obj);
     __ shrptr(tmp3, 6);
     __ movptr(tmp2, ALLOC_BIT_BASE_ADDRESS);
     __ movb(tmp2, Address(tmp2, tmp3));
     // tmp3 = 1 << ((obj >> 3) & 7)
-    //   1. tmp4 = (obj >> 3) & 7
-    __ movptr(tmp4, obj);
-    __ shrptr(tmp4, 3);
-    __ andptr(tmp4, 7);
-    //   2. tmp4 = 1 << tmp4
-    __ movptr(tmp3, rcx);
-    __ movl(rcx, tmp4);
-    __ movptr(tmp4, 1);
-    __ shlptr(tmp4);
-    __ movptr(rcx, tmp3);
-    // tmp2 = tmp2 | tmp4
-    __ orptr(tmp2, tmp4);
+    //   1. rcx = (obj >> 3) & 7
+    __ movptr(rcx, obj);
```

Comment:
This destroys the value in rcx. Do we know that we can safely destroy the value in rcx?

Reply:
Yes. rcx is used to pass the argument to a function, and should be saved by the caller if necessary.

Reply:
It is not related with whether rcx is caller saved or not. But on second thought, it is probably fine, as in this code generation pass, we are still using temporary registers. That suggests the register allocation is done after this pass. We should be fine for using physical registers and the register allocator will deal with it.